src/client-side-metrics/metrics-tracer-factory.ts

```diff
@@ -0,0 +1,600 @@
+   * @returns {Attributes} An object containing the attributes for operation latency metrics.
+   */
+  private getOperationLatencyAttributes(
+    projectId: string,
+    finalOperationStatus: string,
+    streamOperation?: string
+  ): Attributes {
+    return Object.assign(
+      {
+        finalOperationStatus: finalOperationStatus,
+        streamingOperation: streamOperation,
+      },
+      this.getBasicAttributes(projectId)
+    );
+  }
+
+  /**
+   * Assembles the attributes for final operation metrics. These attributes provide
+   * context about the Bigtable environment and the operation being performed.
+   * @param {string} projectId The Google Cloud project ID.
+   * @param {string} finalOperationStatus The final status of the operation.
+   * @returns {Attributes} An object containing the attributes for final operation metrics.
+   */
+  private getFinalOpAttributes(
+    projectId: string,
+    finalOperationStatus: string
+  ): Attributes {
+    return Object.assign(
+      {
+        finalOperationStatus: finalOperationStatus,
+      },
+      this.getBasicAttributes(projectId)
+    );
+  }
+
+  /**
+   * Assembles the attributes for attempt metrics. These attributes provide context
+   * about the Bigtable environment, the operation being performed, and the status of the attempt.
+   * Includes whether the operation was a streaming operation or not.
+   * @param {string} projectId The Google Cloud project ID.
+   * @param {string} attemptStatus The status of the attempt.
+   * @param {string} streamingOperation Whether the operation was a streaming operation or not.
+   * @returns {Attributes} An object containing the attributes for attempt metrics.
+   */
+  private getAttemptAttributes(
+    projectId: string,
+    attemptStatus: string,
+    streamingOperation: string
+  ) {
+    return Object.assign(
+      {
+        attemptStatus: attemptStatus,
+        streamingOperation: streamingOperation,
+      },
+      this.getBasicAttributes(projectId)
+    );
+  }
+
+  /**
+   * Assembles the attributes for attempt status metrics. These attributes provide context
+   * about the Bigtable environment and the operation being performed.
+   * @param {string} projectId The Google Cloud project ID.
+   * @param {string} attemptStatus The status of the attempt.
+   * @returns {Attributes} An object containing the attributes for attempt status metrics.
+   */
+  private getAttemptStatusAttributes(projectId: string, attemptStatus: string) {
+    return Object.assign(
+      {
+        attemptStatus: attemptStatus,
+      },
+      this.getBasicAttributes(projectId)
+    );
+  }
+
+  /**
+   * Called when the operation starts. Records the start time.
+   */
+  onOperationStart() {
+    this.operationStartTime = this.dateProvider.getDate();
+  }
+
+  /**
+   * Called after the client reads a row. Records application blocking latencies.
+   */
+  onRead() {
```

Comment:
Did you do any benchmarking on this one? ReadRows is very performance sensitive. At least in Python, we have to be very thoughtful about literally every line

Reply:
I have now decided to leave this one out for now and include it in a separate PR. The logistics of making this work for the application blocking latencies are rather complicated so I would prefer to prioritize it last.